### PR TITLE
Reservoir separate transforms

### PR DESCRIPTION
**Pull request by AnnaKwa**
This PR adds the ability to load separate transforms for input/output/hybrid variable sets (or fit different standard normalization transforms if not specified). These are organized in a new `TransformerGroup` class that replaces the single `autoencoder` attribute of the reservoir models. 

If the output variable list is different from the input variable list, the training script will use the output transformer when preparing the paired output set.

Added public API:
- The previous single `autoencoder_path` parameter in the `ReservoirTrainingConfig` is now replaced by a `TransformerConfig` which contains optional paths for input, output, and hybrid transformers

Refactored public API:
- Aforementioned API change breaks old configs
- Breaks old model loading, since models now dump and load a `TransformerGroup` instead of a single autoencoder 

 

- [ x] Tests added

### CODE REVIEW

File: external/fv3fit/fv3fit/reservoir/train.py [hunk 1]
```diff
@@ -92,22 +116,32 @@ def train_reservoir_model(
             variables=hyperparameters.input_variables,
             batch_data=batch_data,
             rank_divider=rank_divider,
-            autoencoder=autoencoder,
+            autoencoder=transformers.input,
         )
+        # If the output variables differ from inputs, use the transformer specific
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/reservoir/train.py:121]: I think we need separate arguments for X and y rank divider and transformers with the processing function to simplify the training code.

**Review by frodre — CHANGES_REQUESTED** [review]
Thanks @AnnaKwa, I think there's a few changes necessary before this one is ready.

One essential change is that the transformers may have different `z_feature_size` dimensions, so the rank divider may not work for different transformer outputs due to the shape checking.  Perhaps adding a method to the rank divider that would generate a new rank divider object  (for hybrid/outputs) used for batch processing based on the `n_latent_dims` of the transformer.  Have a divider for inputs/outputs/(maybe hybrids).  Along the lines of this [method](https://github.com/ai2cm/fv3net/blob/b12a2d00bd74cc35c94b3b906dc9c374da1f3f95/external/fv3fit/fv3fit/reservoir/domain2.py#L127).  On one hand it's good to have rigorous checks that the shapes line up, but again trailing features are making things more difficult at implementation time.  (Which I believe was the motivation for removing all `z` info in one of your earlier PRs)

Additionally, the current structure of transformer groups was a bit hard to follow because there are delayed decisions being made about whether a transformer is reused or not.  I would rather see something where we have the training set up which transformers are used for what variable group, but once instantiated and saved, the group attributes will always point to a transformer.  You can use store some context in the class so saving/loading doesn't duplicate, but it's really not that resource intensive to just save everything.  That would remove logic checks about which one to use outside of that group class and the training init function.  Let me know what you think, I'm very open to context or suggestions on this.

File: external/fv3fit/fv3fit/reservoir/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/train.py:121]: I opted to simplify the function formerly known as `process_batch_Xy_data` to just operate on a single set of variables, and use the overlap of the rank_divider provided to determine whether to trim halo points.

File: external/fv3fit/fv3fit/reservoir/train.py [hunk 2]
```diff
@@ -37,6 +37,41 @@ def _add_input_noise(arr: np.ndarray, stddev: float) -> np.ndarray:
...
+        )
+
+    # If output transformer not specified and output_variables != input_variables,
+    # create a separate standard norm transform
+    if "output" not in transformers and (
+        hyperparameters.output_variables != hyperparameters.input_variables
+    ):
+        transformers["output"] = get_standard_normalizing_transformer(
+            hyperparameters.output_variables, sample_batch
+        )
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/train.py:62]: Changed the `TransformerGroup` to require the `output` arg, and still have an optional `hybrid` arg. All logic about using the input transformer in place of another is removed from the `TransformerGroup`.

File: external/fv3fit/fv3fit/reservoir/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/train.py:121]: To allow this function to work on input/output/hybrid data individually, it was necessary to add another arg `trim_halo`. I found this to be the clearest way to specify whether the halo points needed to be excluded. Another arg is not ideal, but IMO more clear than how it was before where first/second items returned were with/without overlap points.

File: external/fv3fit/fv3fit/reservoir/train.py [hunk 3]
```diff
@@ -37,6 +36,44 @@ def _add_input_noise(arr: np.ndarray, stddev: float) -> np.ndarray:
...
+    else:
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/train.py:63]: If `output` is in `transformers`, should defer to the loaded transformer.   Only set the output transformer to input if nothing specified and variables sets are equivalent.  Can be fixed by moving the conditional on variable sets into the `not in transformers` block.

File: external/fv3fit/fv3fit/reservoir/train.py [hunk 4]
```diff
@@ -37,6 +36,44 @@ def _add_input_noise(arr: np.ndarray, stddev: float) -> np.ndarray:
...
+        transformers["output"] = transformers["input"]
+
+    # If hybrid variables transformer not specified, and hybrid variables are defined,
+    # create a separate standard norm transform
+    if "hybrid" not in transformers and hyperparameters.hybrid_variables is not None:
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/train.py:68]: Same situation as `output` transformer above.  Move the `hybrid_variables` check into the `not in transformers` block.

File: external/fv3fit/fv3fit/reservoir/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/reservoir/train.py:123]: Awesome, very good solution to the issue!

File: external/fv3fit/fv3fit/reservoir/transformers/transformer.py [hunk 5]
```diff
@@ -29,6 +30,11 @@ def decode(self, x: ArrayLike) -> Sequence[ArrayLike]:
         pass
 
 
+class Transformer(BaseTransformer, Reloadable):
+    def __init__(self, **kwargs):
+        self.super().__init__(**kwargs)
+
+
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/transformers/transformer.py:37]: `Reloadable` is now redundant in the `DoNothingAutoencoder` subclass instance and can be removed.

File: external/fv3fit/fv3fit/reservoir/utils.py [hunk 6]
```diff
@@ -58,25 +64,42 @@ def get_ordered_X(X: Mapping[str, tf.Tensor], variables: Iterable[str]):
...
-) -> Tuple[np.ndarray, np.ndarray]:
-    """ Convert physical state to corresponding reservoir hidden state,
+    autoencoder: Optional[Transformer],
+    trim_halo: bool,
+):
+    """ Converts physical state to latent state
     and reshape data into the format used in training.
+    The rank divider provided includes the full overlap, since
+    the data it is operating on includes all halo points.
+    The rank
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/utils.py:78]: Leftover in docstring

**Review by frodre — APPROVED** [review]
Looks good!  Thanks for the updates, I think the notion of which encoders are used and when is very clear now.  One minor thing to adjust in the training code that sets up the transform group, but otherwise ready to go.
